Changed hydrateFrom to accept append flag and removed appendFrom method

smolinari opened the pull request:
Not ready to merge!

Just a start. The easy part.

Scott

chrismichaels84 commented:
Looks like a great start. Thanks for the help.

chrismichaels84 commented on src/Bags/FileBag.php:
```diff
@@ -0,0 +1,73 @@
+     * @param $arrayOfSplFileInfoObjects
+     * @throws BadFileInfoObjectException
+     */
+    private function initialize($arrayOfSplFileInfoObjects)
```
I would typecast this `initialize(array $splFileInfoObjects)`

smolinari replied:
Understood.

Scott

chrismichaels84 commented on src/Bags/FileBag.php:
```diff
@@ -0,0 +1,73 @@
+     */
+    protected function isSplFileInfoObject($object)
+    {
+        return $object instanceof  \SplFileInfo ? true : false;
```
You can actual just `return $object instanceof  \SplFileInfo`

smolinari replied:
Oh. Yup. Very well.

Scott

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,181 @@
+        }
+        foreach($mimeTypes as $type)
+        {
+            $this->decoders[$type] = &$decoder;
```
I don't think this needs to be passed by reference.

smolinari replied:
Ah, I wasn't sure about that. I'll remove the reference. I was thinking, no matter how many mime types are defined, they should all still point to the same decoder object.

Scott

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,181 @@
+            $this->fileBag = $tempFileBag->getAllFileInfoObjects();
+            return;
+        }
+        throw new BadFileDataException('The attempt at adding files to the file loader failed, due to the files not
```
Good, clear error message.

smolinari replied:
Cool. I did something right. LOL! :smile: 

Scott

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,181 @@
+            }else{
+                $this->unsupportedFiles[] = $file;
+            }
+        }
```
This is a really great check

chrismichaels84 commented on src/Traits/LoadsFileTrait.php:
```diff
@@ -0,0 +1,7 @@
+<?php
+namespace Michaels\Manager\Traits;
+
+trait LoadsFileTrait
```
I don't think `addFiles()` belongs here. All that really belongs here is a `load` method. See the end comment for more on that.

chrismichaels84 commented:
This is a summary of my thoughts for commit 60650fa.

I want to say that there is nothing wrong with the way you have set this up. There were two ways to go about this: embedding `Manager` into the `FileLoader` and using `FileLoader` as a dependency inside `Manager`. Both are completely valid ways of solving the problem. In order to fit into the workflow of the rest of the package, I would prefer `FileLoader` as a dependency of `Manager` instead of embedding the `Manager` instance inside of `FileLoader`.

From a quick scan, I don't think that's the way this works. I'm not sure if this is the way you've set it up.

Now I could be wrong if the final user would be able to do something like this:
```php
$manager = new Manager([ // Assuming this class uses LoadsFilesTrait
    'some' => [
         'starting' => 'data'
    ]
]);

$manager->add('some.more.starting', 'data');

$manager->addFileDecoder($iniFileDecoder);
$manager->load([
    $splJsonFile, $splYamlFile, $splPhpFile, $splIniFile
], true); // true or false to append versus reset data. I'm not sure which should be default. Thoughts?
```

What you have created could do this wonderfully if `LoadsFilesTrait::load()` sets up the FileLoader, adds the current `Manager` instance to the `FileLoader`, passes in any registered decoders, passes or creates a `FileBag` into that `FileLoader`, and then updates the `Manager` that was passed.

Personally, I would tweak your current implementation so that, instead of actually adding data to `Manager`, `FileLoader' simply returns an array of data from all the files that were loaded, and then `LoadsFilesTrait::load()` actually appends or resets the `Manager`

All that said, I don't know that I would worry about that at the moment. You've got a great start and this is solid code. I would finish it up, show me the end-use-case api (something similar to the above) and we can see how to tweak it. Or, I'm more than happy to tweak it to fit with the workflow I have in mind.

In any case, this is solid code, and I don't think you should be as hard on yourself with noob mistakes. I'm sure I will have some more minor suggestions once you are further along, but I don't like to make comments on that kind of stuff until a pr is more-less finished.

Does all that make sense?

smolinari commented:
Yes, all makes great sense. I was aware my direction wasn't quite right. Your explanation above helps and I have a plan. Only need some minor changes and now I actually understand where the trait can come in handy. :+1: 

I am hard on myself, because I am a noob. Seriously, I am. I've been looking at PHP code for the last 13 years more from a service rep position. I've self-taught myself PHP and now the past year and a half or so OOP programming. These are my first serious attempts at actually producing finished code. Yes, I've never really coded anything of any value before. 

What I like about doing this is, it is code I need for my project, which I realize is a mammoth project, but it also means I've started on it. These is my first baby steps. :smile: And, at the same time, because we are helping each other  and  the support you are giving me, and I get some of your experience to boot to help my knowledge level. A win, win situation.

I can't wait to get Spider finished and start to use it. It will be at the core of my project.   

Scott

chrismichaels84 commented:
Glad to hear it. I have a major deadline coming up on Oct 15th, but after that, I will be able to dedicate mostly the entire rest of the year to getting Spider on its feet. Let me know if you need anything else.

smolinari commented:
We now have the API additions.

$manager->loadFiles($files, $append)

$manager->addDecoder(\DecoderInterface $decoder)

Things to do still (so I don't forget. :smile:)

 Create the YAML decoder with tests
 Create the PHP decoder with tests
 Create some custom decoder and test its addition and decoding properly (suggestions on what kind are welcome).
 Do tests for appending with file data
Any other tests?

Scott

smolinari commented:
Finished the Yaml Decoder.

Scott

smolinari commented:
I am not too sure about how the PHP decoder works. It isn't really modular, since I have to check for a PHP extension in the FileLoader, to send the right data (`include()` instead of `file_get_contents()`) to the `decode()` method, which just sends the data back. I tried to make it work sending the contents of the file, but couldn't. 

Scott

smolinari commented:
Ok. Custom decoder also "added" (both in testing and as a new XML decoder). I've also updated the docs. Hope it is ok. :smile: 

Scott

smolinari commented:
Hmm...I don't understand why the travis attempt at running the tests doesn't find the CustomXMLDecoder.php class. Do you have any idea Michael?

Scott

chrismichaels84 commented:
I just got back into town. Let me take a look at this today.

smolinari commented:
Hope all is well with you and it is more a problem with my code. But, I hope it didn't make you so sick you couldn't return to it. LOL! :smile: 

Scott

chrismichaels84 commented:
Not at all. I've had some personal emergencies. I promise this is a priority and I will get to it as fast as I can. I also have some updates for Spider. Thank you for everything

smolinari commented:
Oh, personal emergencies doesn't sound too good. No worries. Take care of yourself first. 

Scott

chrismichaels84 commented:
Aha! I'm finally back! And I've taken a look at the travis test. It passes on my machine and I was stumped for a minute until my IDE told me the problem.

The classname for `Michaels\Manager\Decoders\CustomXmlDecoder` does not match the filename `/src/Decoders/CustomXMLDecoder.php`. The filename has an all caps XML where the class has Xml. I know its small, but that is causing composer not to load that file, thus the class isn't available. Different OS's and servers handle filenames differently, so it MAY pass on some machines but not others.

That simple fix should cause Travis to pass.

Now, let me look through the rest real quick, but I think this seems ready to merge into develop. Was there anything else you wanted to do before it merges?

chrismichaels84 commented:
I am also going through real fast for a few inline comments. Very small. Mostly to make the style and whatnot fit the rest of the project.

chrismichaels84 commented on src/Traits/LoadsFilesTrait.php:
```diff
@@ -0,0 +1,57 @@
+use Michaels\Manager\FileLoader;
+
+trait LoadsFilesTrait
+{
```
be sure to `use Michaels\Manager\Traits\DependsOnManagesItemsTrait` which only ensures that the Trait isn't being used without `ManagesItemsTrait`. See `ChainsNestedItemsTrait` for an example.

chrismichaels84 replied:
Actually, for now, you can drop that thought. I didn't notice that you actually add the manager as a dependency in `LoadsFilesTrait`. I like that approach, but it is unique in the system, so I'm not sure if it would be better to leave it or to rework it a it so that it just manipulates the current `ManagesItemsTrait` directly, simply for consistency in the library. I'll stew over that and probably decide after the merge. Right now, if it ain't broke, don't fix it! lol

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,208 @@
+
+            include_once(__DIR__.'/Decoders/'.$decoderClass.'.php');
+            $nameSpace = "Michaels\\Manager\\Decoders\\";
+            $fqcn = $nameSpace . $decoderClass;
```
With IDE's, its simple enough to use more complete variable names. Maybe even `fullClass`.

Also, we use Composer to autoload files, so we don't need to include that file manually. Just remove 200

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,208 @@
+        $this->fileBag = array();
+        $this->supportedMimeTypes = array();
+        $this->unsupportedFiles = array();
+        $this->dataToDecode = array();
```
Please declare this property at the top of the class.

chrismichaels84 commented on src/FileLoader.php:
```diff
@@ -0,0 +1,208 @@
+class FileLoader
+{
+
+    private $supportedMimeTypes = [];
```
Please docblock these. Also, I will make them `protected` when I merge.

chrismichaels84 commented:
Some specific line-note thoughts. See line notes in files for more details:
  - [ ] Remove line 200 in FileLoader.php
  - [ ] I will probably refactor `FileLoader` NOT to depend on a `ManagesItems` instance and move `hydrateManger()` back into Manager. Simply because that would make it more consistent with the rest of the library and allow people to use the `FileLoader` class independently of `Manager` if they want to. But, I can take care of that when I merge other branches into `develop` before I release.
  - [ ] Please docblock all of the class properties and make sure to declare them all. (FileLoader properties and line 134 `dataToDecode` not declared)
  - [ ] Let's move `tests/bags/FileBagTestTrait.php` into a new directory called `tests/Utilities` to keep it separate from the actual tests.

Some general thoughts:
  - [ ] This library adhers to [PSR-2 coding style](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) which isn't a big deal because my IDE and Scrutinzer can both auto format the code to that style. But a couple things I notices are that opening brackets for `foreach` and `if` statements go on the same line. Also, always be sure to leave exactly one blank line at the end of text files. Some operating systems and servers will get buggy without it. (IDK why)
  - [ ] I like to use `protected` instead of `private`, but as we've discussed that's largely personal preference. To me, a library should be extendable by default.

Other than those minor things, all looks great! It's good work with clean tests and clear documentation. I'm happy to merge whenever you feel you are finished. Thank you for your help.

Also, just curious, which IDE or text editor do you use?

smolinari commented:
I use PHPStorm9. I'll get to the rest of the comments, when I get back home. 

Scott

smolinari commented:
>The classname for Michaels\Manager\Decoders\CustomXmlDecoder does not match the filename /src/Decoders/CustomXMLDecoder.php.

Strange, on my computer both the filename and the class have Xml in them. 

>With IDE's, its simple enough to use more complete variable names. Maybe even fullClass.
> Also, we use Composer to autoload files, so we don't need to include that file manually. Just remove 200

Corrected.

> Please declare this property at the top of the class.

Corrected. I missed the warning from my IDE about it missing too.

> Please declare this property at the top of the class.
> Please docblock these. Also, I will make them protected when I merge.

Corrected.

> Let's move tests/bags/FileBagTestTrait.php into a new directory called tests/Utilities to keep it separate from the actual tests.

Done.

> This library adhers to PSR-2 coding style which isn't a big deal because my IDE and Scrutinzer can both auto format the code to that style. But a couple things I notices are that opening brackets for foreach and if statements go on the same line. Also, always be sure to leave exactly one blank line at the end of text files. Some operating systems and servers will get buggy without it. (IDK why)

I thought I was following PSR-2, but I guess I missed some formatting here and there. As I mentioned, I am not a practiced coder. So, learning, learning, learning. I think I corrected my mistakes. I've added the blank lines to my additions too, however, I also noticed your files are missing the blank line. :smile: 

> I like to use protected instead of private...

I can understand your protected over private preference. Though, I thought a library should be an encapsulated piece of code and not really up for extension. If the library is used in a framework or a larger component (like Spider), then I could fully understand that code (Spider's code) being extendable. Still, what do I know? I am not even a junior coder. LOL!

The only other concern I have is what Dylan said here about stubbing: https://github.com/chrismichaels84/data-manager/issues/29

I wish I understood better, because I never got it done. He correctly noted that I was doing functional testing and not unit testing (I believe that was the issue). 

At any rate. I am, for what it is worth, happy my code works and the code isn't total crap. :smile:

The commit is up. 

Scott

smolinari commented:
Very strange. I don't know why, but the tests were all passing on my test server, even though I forgot to update the use statements in the test classes. 

Scott

smolinari commented:
Hmm....now only the HHVM test failing and I have no clue why.

Scott

PommeVerte commented:
Shot in the dark but perhaps something to do with HHVM's version of `SPLFileInfo`, or how it handles `error_reporting()`? Dunno just a wild guess.

chrismichaels84 commented:
I am going to merge this into develop so that I can track down the bug more easily without having to do PRs against a PR. Everything else looks up to speed.

chrismichaels84 commented:
Well, I tracked down the problem. When using `assertEquals()` in HHVM, it appears to be much more strict. Because `FileLoader` `unshifts` the values onto the array, they are actually in reverse order inside the `FileLoader`. In PHP, the order of an array doesn't seem to affect its equalness, but in HHVM, it does. Working on a fix now.

chrismichaels84 commented:
Alright, I fixed it! I also learned some stuff, so I thought I'd pass it along.

  * HHVM's `assertEquals` is much more strict for whatever reason.
  * The files were being added to the fileBag in reverse order, which was the problem
  * I changed `array_shift` to `array_push` which cleared up the error, but brought out another bug
  * one of the fixture config.yaml files was a little different than the other config files. Since that file was the last to be added, it was testing against that and throwing some new errors. I corrected it and allis well.
  * On that note, we should look at reworking those tests. ATM, each config file that is added overwrites the previous one because there is no namespacing, so (basically) only the last to be added is being tested.
  * I think that's a design flaw on my end. It should probably add file data under a namespace of the file's name. I want to look around at what others do, but I think that is typical.
  * Right now, if two files both have the same key, the second one just overwrites the first.

I also refactored and cleaned up a bit. Nothing major. I moved hydrate out into the `ManagesItems` trait itself and just had the `FileLoader` return array data instead of hydrating the manager. Seemed more SRP to me, and this way it is possible for someone to use that class independently if they wish.

I know its a library, but to me EVERYTHING should be extendable and flexible. I simply don't know the future and have myself found that I often use a library just for one of its component classes, ignoring the libraries actual function, lol.

In any case, this is good work and has landed safely in `develop` and will be released (maybe experimentally) in the next version. Hopefully today.

smolinari commented:
Sounds good and thanks for the feedback. To be honest, I wasn't sure about which to use, `array_shift` or `array_push` and it is interesting HHVM helps "push" us to `array_push`. LOL!  I'll take a deeper dive into the code a bit later to look at (and learn from) the changes. Sorry, my work meant so much extra work for you. 

Scott

chrismichaels84 commented:
All code needs extra work. Yours didn't need any more than mine would have if I did it.
